lib/msal-browser/src/cache/BrowserStorage.ts

```diff
@@ -0,0 +1,239 @@
+import { BrowserConstants } from "../utils/BrowserConstants";
+import { ErrorCacheKeys } from "msal-common/dist/utils/Constants";
+
+export class BrowserStorage implements ICacheStorage {
+
+    private cacheConfig: CacheOptions;
+    private windowStorage: Storage;
+
+    private clientId: string;
+    private rollbackEnabled: boolean;
+
+    constructor(clientId: string, cacheConfig: CacheOptions) {
+        if (!window) {
+            throw BrowserAuthError.createNoWindowObjectError();
+        }
+
+        const storageSupported = typeof window[cacheConfig.cacheLocation] !== "undefined" && window[cacheConfig.cacheLocation] !== null;
+        if (!storageSupported) {
+            throw BrowserConfigurationAuthError.createStorageNotSupportedError();
+        }
+
+        this.cacheConfig = cacheConfig;
+        this.windowStorage = window[this.cacheConfig.cacheLocation];
+        this.clientId = clientId;
+
+        // This is hardcoded to true for now, we will roll this out as a configurable option in the future.
+        this.rollbackEnabled = true;
+
+        this.migrateCacheEntries();
+    }
+
+    /**
+     * Support roll back to old cache schema until the next major release: true by default now
+     * @param storeAuthStateInCookie
+     */
+    private migrateCacheEntries() {
+
+        const idTokenKey = `${Constants.CACHE_PREFIX}.${PersistentCacheKeys.ID_TOKEN}`;
+        const clientInfoKey = `${Constants.CACHE_PREFIX}.${PersistentCacheKeys.CLIENT_INFO}`;
+        const errorKey = `${Constants.CACHE_PREFIX}.${ErrorCacheKeys.ERROR}`;
+        const errorDescKey = `${Constants.CACHE_PREFIX}.${ErrorCacheKeys.ERROR_DESC}`;
+
+        const idTokenValue = this.getItem(idTokenKey);
+        const clientInfoValue = this.getItem(clientInfoKey);
+        const errorValue = this.getItem(errorKey);
+        const errorDescValue = this.getItem(errorDescKey);
+
+        const values = [idTokenValue, clientInfoValue, errorValue, errorDescValue];
+        const keysToMigrate = [PersistentCacheKeys.ID_TOKEN, PersistentCacheKeys.CLIENT_INFO, ErrorCacheKeys.ERROR, ErrorCacheKeys.ERROR_DESC];
+
+        keysToMigrate.forEach((cacheKey, index) => this.createCacheEntry(cacheKey, values[index]));
+    }
+
+    /**
+     * Utility function to help with rollback keys
+     * @param newKey
+     * @param value
+     * @param storeAuthStateInCookie
+     */
+    private createCacheEntry(newKey: string, value: string) {
+        if (newKey && value) {
+            this.setItem(newKey, value);
+        }
+    }
+
+    /**
+     * Prepend msal.<client-id> to each key; Skip for any JSON object as Key (defined schemas do not need the key appended: AccessToken Keys or the upcoming schema)
+     * @param key
+     * @param addInstanceId
+     */
+    private generateCacheKey(key: string, addInstanceId: boolean): string {
+        try {
+            // Defined schemas do not need the key appended
+            JSON.parse(key);
```

Comment:
I might create a small helper function called `validateKey` that runs `JSON.parse`, to help make it clear what is going on here.

Reply:
Done.